velox/exec/Unnest.cpp

```diff
@@ -111,20 +112,61 @@ RowVectorPtr Unnest::getOutput() {
   }
 
   const auto size = input_->size();
-  const auto maxOutputSize = outputBatchRows();
 
   // Limit the number of input rows to keep output batch size within
-  // 'maxOutputSize' if possible. Process each input row fully. Do not break
-  // single row's output into multiple batches.
+  // 'maxOutputSize' if possible. Not process each input row fully when single
+  // row's output exceeds maxOutputSize. Single row's output maybe into
+  // multiple batches.
   vector_size_t numInput = 0;
   vector_size_t numElements = 0;
-  for (auto row = nextInputRow_; row < size; ++row) {
-    numElements += rawMaxSizes_[row];
+  vector_size_t partialProcessRowStartSize = -1;
+  vector_size_t firstRowEndSize = -1;
+  // Process first row.
+  if (nextInputRow_ < size) {
+    auto firstRow = nextInputRow_;
+    firstRowEndSize = rawMaxSizes_[firstRow];
+    vector_size_t remainingSize = firstRowEndSize - firstRowStartSize_;
+    if (numElements + remainingSize > maxOutputSize_) {
+      // Single row's output is into multiple batches.
+      // Read the size range from them, not use 0 to rawMaxSizes_[row].
+      firstRowEndSize = firstRowStartSize_ + maxOutputSize_ - numElements;
+      // Process maxOutputSize_ in this getOutput.
+      numElements = maxOutputSize_;
+      partialProcessRowStartSize = firstRowEndSize;
+    } else {
+      // No need to split this row.
+      numElements += remainingSize;
+    }
     ++numInput;
-
-    if (numElements >= maxOutputSize) {
+  }
+  // Not split middle rows.
+  // If there is only 1 row, the end row will not take effect, its startSize is
+  // always 0.
+  vector_size_t endRowEndSize = -1;
+  for (auto row = nextInputRow_ + 1; row < size; ++row) {
+    if (numElements >= maxOutputSize_) {
       break;
     }
+    vector_size_t remainingSize = rawMaxSizes_[row];
+    if (numElements + remainingSize > maxOutputSize_) {
+      // This is the end row.
```

Comment:
This code looks very similar to the logic for processing first row. Wondering if these could be consolidated.